### PR TITLE
[SYCL] add preliminary support for bfloat16 to sycl::vec

### PR DESCRIPTION
**Pull request by cperkinsintel**
This does not yet add support for the math builtins for `sycl::vec<bfloat16>`  . That will come later.

### CODE REVIEW

File: sycl/include/sycl/types.hpp [hunk 1]
```diff
@@ -495,7 +533,12 @@ template <typename Type, int NumElements> class vec {
   template <size_t... Is>
   constexpr vec(const std::array<vec_data_t<DataT>, NumElements> &Arr,
                 std::index_sequence<Is...>)
-      : m_Data{vec_data_t<DataT>(static_cast<DataT>(Arr[Is]))...} {}
+      : m_Data{([&](vec_data_t<DataT> v) constexpr {
+          if constexpr (std::is_same_v<sycl::ext::oneapi::bfloat16, DataT>)
+            return v.value;
+          else
```
Comment (aelovikov-intel) [inline sycl/include/sycl/types.hpp:539]: Why do we need this? Don't we provide conversion operators that would work with the `static_cast`?

**Comment by aelovikov-intel** [comment]
Based on https://reviews.llvm.org/D150913 and https://llvm.org/docs/LangRef.html#floating-point-types I'd think that we should re-implement `bfloat16` support from scratch using "native" clang/llvm support, at least for the device part.

+ @steffenlarsen , @AlexeySachkov , @sergey-semenov , @jbrodman , @elizabethandrews .

File: sycl/include/sycl/detail/generic_type_lists.hpp [hunk 2]
```diff
@@ -91,13 +123,19 @@ using floating_list =
 // geometric floating point types
 using scalar_geo_half_list = type_list<half>;
 
+using scalar_geo_bfloat16_list = type_list<sycl::ext::oneapi::bfloat16>;
```
[the remark this answers is not in the document]
Comment (cperkinsintel) [reply sycl/include/sycl/detail/generic_type_lists.hpp:126]: This PR does not yet bring bfloat16 into the full builtins support.  That _will_ be a separate PR.  But it needs to be added to some of these lists to get sycl::vec<bfloat16> to work for things like simple vector arithmetic.

File: sycl/include/sycl/ext/oneapi/bfloat16.hpp [hunk 3]
```diff
@@ -30,9 +36,31 @@ namespace detail {
...
+using Vec8StorageT = Bfloat16StorageT __attribute__((ext_vector_type(8)));
+using Vec16StorageT = Bfloat16StorageT __attribute__((ext_vector_type(16)));
+#else
+using Vec2StorageT = std::array<Bfloat16StorageT, 2>;
+using Vec3StorageT = std::array<Bfloat16StorageT, 3>;
+using Vec4StorageT = std::array<Bfloat16StorageT, 4>;
+using Vec8StorageT = std::array<Bfloat16StorageT, 8>;
+using Vec16StorageT = std::array<Bfloat16StorageT, 16>;
+#endif
+} // namespace bf16
```
[the remark this answers is not in the document]
Comment (cperkinsintel) [reply sycl/include/sycl/ext/oneapi/bfloat16.hpp:55]: Half and the cl types have similar tables.  Essentially, `sycl::vec` gets mapped to `ext_vector_type` on the device for performance. That is for vector sizes 1, 2, 3, 4, 8 and 16.  So that's what these declarations are for. On the host we use std::array for this type.

File: sycl/include/sycl/ext/oneapi/bfloat16.hpp [hunk 4]
```diff
@@ -42,13 +70,21 @@ class bfloat16 {
...
 private:
   static detail::Bfloat16StorageT from_float_fallback(const float &a) {
+#ifdef __INTEL_PREVIEW_BREAKING_CHANGES
     if (sycl::isnan(a))
       return 0xffc1;
+#else
+    if (detail::float_is_nan(a))
+      return 0xffc1;
+#endif
+
```
[the remark this answers is not in the document]
Comment (cperkinsintel) [reply sycl/include/sycl/ext/oneapi/bfloat16.hpp:87]: Right now bfloat16.hpp includes builtins.hpp (just to get sycl::isnan).  But as we add bfloat16.hpp into vec, we need to reverse that header inclusion and have builtins include bfloat16.hpp.   So this is a consequence of avoiding that inclusion cycle.

File: sycl/include/sycl/ext/oneapi/bfloat16.hpp [hunk 5]
```diff
@@ -199,6 +243,19 @@ class bfloat16 {
...
+    O << static_cast<float>(rhs);
+    return O;
+  }
+
+  inline friend std::istream &operator>>(std::istream &I, bfloat16 &rhs) {
+    float ValFloat = 0.0f;
+    I >> ValFloat;
+    rhs = ValFloat;
+    return I;
+  }
```
[the remark this answers is not in the document]
Comment (cperkinsintel) [reply sycl/include/sycl/ext/oneapi/bfloat16.hpp:258]: sycl::half has it, so it seemed like a we should have parity.  Our tests that use FileCheck can't be extended for `vec<bfloat16>` support without it.  Presently, this PR is the only user of these ops, but that will definitely change as time passes.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-unified-utils.hpp [hunk 6]
```diff
@@ -7,6 +7,10 @@
 // ===--------------------------------------------------------------------=== //
 
 #pragma once
+
+#include <CL/__spirv/spirv_types.hpp> // __spv namespace
+#include <optional>                   // std::optional
```
Comment (dkhaldi) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-unified-utils.hpp:12]: what is the reason behind this change?

Comment (cperkinsintel) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-unified-utils.hpp:12]: This matrix code was getting std::optional and spv namespace by virtue of an inclusion to bfloat16.hpp which in turn included the math builtins which included those things. 
But now bfloat16.hpp is a dependency OF the math builtins and itself no longer includes them (otherwise cyclic dependency). So this matrix file now needs to include the resources it uses directly.

File: sycl/include/sycl/types.hpp [hunk 7]
```diff
@@ -142,20 +144,53 @@ using select_apply_cl_t = std::conditional_t<
...
+  static constexpr RetType set(bool value) { return value; }
+};
+
+template <> struct vec_helper<sycl::ext::oneapi::bfloat16> {
+  using RetType = sycl::ext::oneapi::bfloat16;
+  using BFloat16StorageT = sycl::ext::oneapi::detail::Bfloat16StorageT;
+  static constexpr RetType get(BFloat16StorageT value) {
+    // awkward workaround. sycl::bit_cast isn't constexpr on Win.
+    // C++20 will give us both std::bit_cast and constexpr reinterpet for void*
+    // but neither available yet.
```
[the remark this answers is not in the document]
Comment (cperkinsintel) [reply sycl/include/sycl/types.hpp:162]: The "needs compiler magic" from cppreference is because the possible implementation that is provided there using memcpy which is not constexpr.

File: sycl/include/sycl/types.hpp (hunk 1 above)
Comment (cperkinsintel) [reply sycl/include/sycl/types.hpp:535]: The storage type for bfloat16 is `uint16_t` and that's the data behind the `m_Data` but the `DataT` is bfloat16. There is, intentionally, no direct conversion between those types. If you cast, the compiler will insert a call to the bfloat16 float operator() for the conversion, which is unnecessary and, in many cases, will mess up the value.  So there a couple of places the vec<bfloat16> needs to be careful.

**Comment by AlexeySachkov** [comment]
> Based on https://reviews.llvm.org/D150913 and https://llvm.org/docs/LangRef.html#floating-point-types I'd think that we should re-implement `bfloat16` support from scratch using "native" clang/llvm support, at least for the device part.
> 
> * @steffenlarsen , @AlexeySachkov , @sergey-semenov , @jbrodman , @elizabethandrews .

I suppose that this is something we should consider doing, but it would definitely be a more complicated change. With custom C++-level wrapper class, we can translate it into SPIR-V just fine, since it is only a wrapper on top of `uint16_t`. When it is turned into a native LLVM IR type, then SPIRV-LLVM-Translator will have to be updated as well to recognize that type, which will require a SPIR-V extension for native bfloat16 and support for it in backends

File: sycl/include/sycl/stream.hpp [hunk 8]
```diff
@@ -346,6 +347,26 @@ checkForInfNan(char *Buf, T Val) {
   return 0;
 }
 
+template <typename T>
+inline typename std::enable_if_t<std::is_same_v<T, ext::oneapi::bfloat16>,
+                                 unsigned>
+checkForInfNan(char *Buf, T Val) {
```
[the remark this answers is not in the document]
Comment (cperkinsintel) [reply sycl/include/sycl/stream.hpp:353]: Good point - these three overloads of `checkForInfNan`  in stream.hpp perhaps didn't all need to be templates. But that's how it is, including the already existing one for sycl::half. So it makes sense to just do the same for bfloat16.  As much as possible this PR adds bfloat16 support by riding the coattails of sycl::half, or mimicking it.  
I can change them, if you want.

File: sycl/include/sycl/stream.hpp [hunk 9]
```diff
@@ -346,6 +347,26 @@ checkForInfNan(char *Buf, T Val) {
...
+  if (Val != Val)
+    return append(Buf, "nan");
+
+  // Extract the sign from the bits
+  const uint16_t Sign = reinterpret_cast<uint16_t &>(Val) & 0x8000;
```
[the remark this answers is not in the document]
Comment (cperkinsintel) [reply sycl/include/sycl/stream.hpp]: done

File: sycl/include/sycl/types.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (cperkinsintel) [reply sycl/include/sycl/types.hpp]: With the latest changes to `sycl::bit_cast`, it is now constexpr on Windows. The only holdout where `sycl::bit_cast` isn't constexpr is on older GCC. Therefore, I have changed this code to now use `sycl::bit_cast` in the cases where a constexpr version is available.  So this awkward union treatment is relegated to the corner of someone both using an older GCC and also using `sycl::vec<bfloat16>`.

**Comment by dkhaldi** [comment]
Does this change enable sub_group::load and sub_group::store (https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/sub_group.hpp#L71) to work on bfloat16 type?
If yes, it will be good to add an example of these two instructions in one of the tests you added.

**Comment by cperkinsintel** [comment]
@dkhaldi  - I did a quick test of using sub_groups to .load() data from one buffer and modify and .store() it into another, and this works with bfloat16 in the _current_ source base.  So, if you are asking if this PR would enable that behavior, the answer is "no" - it's already present and working. AFAICT.  

But maybe I'm misunderstanding your inquiry?

I also checked using sub_groups to .load()/.store() for `sycl::vec<bfloat16>` and that now compiles and works with this PR, whereas before it, it would not (obviously).   Are you sure we need testing for this?  I strongly suspect the contexts of your inquiry and my answer are not the same.

**Comment by dkhaldi** [comment]
> I also checked using sub_groups to .load()/.store() for `sycl::vec<bfloat16>` and that now compiles and works with this PR, whereas before it, it would not (obviously). Are you sure we need testing for this? I strongly suspect the contexts of your inquiry and my answer are not the same.

Sorry I was not specific, Yes, I meant this latter. sub_group::load/store on `sycl::vec` of type bfloat16.
Good to know that your PR enables that.
I think it will be good to add that in the test since your PR enables it.

File: sycl/include/sycl/detail/generic_type_lists.hpp [hunk 10]
```diff
@@ -131,12 +163,15 @@ using geo_list = tl_append<scalar_geo_list, vector_geo_list>;
 // cross floating point types
 using cross_half_list = type_list<vec<half, 3>, vec<half, 4>>;
 
+using cross_bfloat16_list = type_list<vec<sycl::ext::oneapi::bfloat16, 3>,
+                                      vec<sycl::ext::oneapi::bfloat16, 4>>;
+
```
Comment (aelovikov-intel) [inline sycl/include/sycl/detail/generic_type_lists.hpp:168]: I don't think this is needed either.

File: sycl/include/sycl/stream.hpp [hunk 11]
```diff
@@ -83,10 +84,10 @@ constexpr size_t MAX_ARRAY_SIZE =
 constexpr unsigned FLUSH_BUF_OFFSET_SIZE = 2;
 
 template <class F, class T = void>
-using EnableIfFP = typename std::enable_if_t<std::is_same_v<F, float> ||
-                                                 std::is_same_v<F, double> ||
-                                                 std::is_same_v<F, half>,
-                                             T>;
+using EnableIfFP = typename std::enable_if_t<
+    std::is_same_v<F, float> || std::is_same_v<F, double> ||
+        std::is_same_v<F, half> || std::is_same_v<F, ext::oneapi::bfloat16>,
```
Comment (aelovikov-intel) [inline sycl/include/sycl/stream.hpp:89]:
```suggestion
    detail::check_type_in_v<F, float, double, half, ext::oneapi::bfloat16>,
```

Might need https://github.com/intel/llvm/pull/12454 to land first.

File: sycl/test-e2e/BFloat16/bfloat16_vec.cpp [hunk 12]
```diff
@@ -0,0 +1,159 @@
+//==------------ bfloat16_vec.cpp - test sycl::vec<bfloat16>----------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: cpu || accelerator
```
Comment (aelovikov-intel) [inline sycl/test-e2e/BFloat16/bfloat16_vec.cpp:9]: Why? If simply not yet supported there, add a comment stating that.

File: sycl/include/sycl/detail/generic_type_lists.hpp (hunk 10 above)
Comment (cperkinsintel) [reply sycl/include/sycl/detail/generic_type_lists.hpp]: removed

File: sycl/include/sycl/detail/generic_type_traits.hpp [hunk 13]
```diff
@@ -539,10 +547,9 @@ using select_cl_scalar_t = std::conditional_t<
     std::is_integral_v<T>, select_cl_scalar_integral_t<T>,
     std::conditional_t<
         std::is_floating_point_v<T>, select_cl_scalar_float_t<T>,
-        // half is a special case: it is implemented differently on
-        // host and device and therefore, might lower to different
-        // types
-        std::conditional_t<is_half_v<T>,
+        // half and bfloat16 are special cases: they are implemented differently
+        // on host and device and therefore might lower to different types
+        std::conditional_t<is_half_or_bf16_v<T>,
```
[the remark this answers is not in the document]
Comment (cperkinsintel) [reply sycl/include/sycl/detail/generic_type_traits.hpp]: good catch, thanks. Fixed.

File: sycl/include/sycl/ext/oneapi/bfloat16.hpp [hunk 14]
```diff
@@ -30,9 +36,31 @@ namespace detail {
...
+using Vec2StorageT = std::array<Bfloat16StorageT, 2>;
+using Vec3StorageT = std::array<Bfloat16StorageT, 3>;
+using Vec4StorageT = std::array<Bfloat16StorageT, 4>;
+using Vec8StorageT = std::array<Bfloat16StorageT, 8>;
+using Vec16StorageT = std::array<Bfloat16StorageT, 16>;
+#endif
+} // namespace bf16
+
+#ifndef __INTEL_PREVIEW_BREAKING_CHANGES
+static bool float_is_nan(float x) { return x != x; }
```
[the remark this answers is not in the document]
Comment (cperkinsintel) [reply sycl/include/sycl/ext/oneapi/bfloat16.hpp]: done

File: sycl/test-e2e/BFloat16/bfloat16_vec.cpp (hunk 12 above)
Comment (cperkinsintel) [reply sycl/test-e2e/BFloat16/bfloat16_vec.cpp:11]: added

**Review by aelovikov-intel — APPROVED** [review]
Looks much better now, but I've had too many iterations reviewing this that some things might have become invisible to my eye. @steffenlarsen , @sergey-semenov , would you mind having a look through it as well?

**Comment by cperkinsintel** [comment]
@intel/dpcpp-tools-reviewers could I get one of you to review?

**Review by dkhaldi — APPROVED** [review]
LGTM

File: sycl/include/sycl/types.hpp [hunk 15]
```diff
@@ -142,20 +144,61 @@ using select_apply_cl_t = std::conditional_t<
 template <typename T> struct vec_helper {
   using RetType = T;
   static constexpr RetType get(T value) { return value; }
+  static constexpr RetType set(T value) { return value; }
```
Comment (steffenlarsen) [inline sycl/include/sycl/types.hpp:147]: Nit; I am not much for the name "set". What does it set? "get" isn't great for that matter, but it sounds more and more like we are mutating something, which we are not.

File: sycl/include/sycl/types.hpp [hunk 16]
```diff
@@ -393,11 +439,7 @@ template <typename Type, int NumElements> class vec {
   }
   template <typename DataT_, typename T>
   static constexpr auto FlattenVecArgHelper(const T &A) {
-#if defined(__INTEL_PREVIEW_BREAKING_CHANGES)
     return std::array<DataT_, 1>{vec_data<DataT_>::get(static_cast<DataT_>(A))};
-#else
-    return std::array<DataT_, 1>{vec_data<DataT_>::get(A)};
-#endif
```
Comment (steffenlarsen) [inline sycl/include/sycl/types.hpp:400]: Is there a reason we can promote this out of preview now? Was it just a mistake to have it under preview before?

Comment (cperkinsintel) [reply sycl/include/sycl/types.hpp:400]: Correct. It was a mistake and shouldn't have been there to begin with.
